src/iter/interleave.rs

```diff
@@ -0,0 +1,286 @@
+        // Find a, b satisfying the following conditions:
+        // (1) 0 < a <= self.i_len
+        // (2) 0 < b <= self.j_len
+        // (3) a + b == index
+        // assert!(index <= self.i_len + self.j_len);
+
+        let odd_offset = |flag| if flag { 0 } else { 1 };
+
+        // desired split
+        let (i_idx, j_idx) = (idx + odd_offset(even || self.flag),
+                              idx + odd_offset(even || !self.flag));
+
+        let (i_split, j_split) = if self.i_len >= i_idx && self.j_len >= j_idx {
+            (i_idx, j_idx)
+        } else if self.i_len >= i_idx {
+            // j too short
+            (index - self.j_len, self.j_len)
+        } else {
+            // i too short
+            (self.i_len, index - self.i_len)
+        };
+
+        let trailing_flag = even == self.flag;
+        let (i_left, i_right) = self.i.split_at(i_split);
+        let (j_left, j_right) = self.j.split_at(j_split);
+
+        (InterleaveProducer::new(
+            i_left,
+            j_left,
+            i_split,
+            j_split,
+            self.flag
+        ), InterleaveProducer::new(
+            i_right,
+            j_right,
+            self.i_len - i_split,
+            self.j_len - j_split,
+            trailing_flag
+        ))
+    }
+}
+
+
+/// Wrapper for Interleave to implement DoubleEndedIterator and
+/// ExactSizeIterator
+pub struct InterleaveSeq<I, J> {
+    i: I,
+    j: J,
+    flag: bool
+}
+
+impl<I, J> Iterator for InterleaveSeq<I, J>
+    where I: Iterator,
+          J: Iterator<Item = I::Item>
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.flag = !self.flag;
+        if self.flag {
+            match self.i.next() {
+                None => self.j.next(),
+                r => r
+            }
+        } else {
+            match self.j.next() {
+                None => self.i.next(),
+                r => r
+            }
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (ih, jh) = (self.i.size_hint(), self.j.size_hint());
+        let min = ih.0.checked_add(jh.0).unwrap_or(usize::MAX);
+        let max = match (ih.1, jh.1) {
+            (Some(x), Some(y)) => x.checked_add(y),
+            _=> None
+        };
+        (min, max)
+    }
+}
+
+// The implementation for DoubleEndedIterator requires
+// ExactSizeIterator to provide `next_back()`. The last element will
+// come from the iterator that runs out last (ie has the most elements
+// in it). If the iterators have the same number of elements, then the
+// last iterator will provide the last element.
+impl<I, J> DoubleEndedIterator for InterleaveSeq<I, J>
+    where I: DoubleEndedIterator + ExactSizeIterator,
+          J: DoubleEndedIterator<Item = I::Item> + ExactSizeIterator<Item = I::Item>
+{
+    #[inline]
+    fn next_back(&mut self) -> Option<I::Item> {
+        if self.i.len() <= self.j.len() {
```

Comment:
I think if they are equal length, then it needs to take the `flag` into account.

Reply:
I agree, done.